File: cpp-bson/Lexer.cpp
#include "Lexer.hpp"
#include <sstream>
#include <regex>
#include <iostream>
#include <algorithm>

Lexer::Lexer(const std::string& content) : content(content) {}

// tokenize
// The main loop that processes the input line by line.
// It handles high-level structure like headers, comments, and indentation.
std::vector<Token> Lexer::tokenize() {
    std::stringstream ss(content);
    std::string line;
    int lineNum = 0;
    bool firstLine = true;

    while (std::getline(ss, line)) {
        lineNum++;
        // Handle Windows-style line endings
        if (!line.empty() && line.back() == '\r') line.pop_back();

        // Header Check: The Cry
        if (firstLine) {
            if (line != "BULBA!") {
                throw std::runtime_error("Status: Fainted");
            }
            tokens.push_back({TOKEN_HEADER, "BULBA!", lineNum, 0});
            firstLine = false;
            continue;
        }

        // Handle comments (Sleep Powder)
        // We strip out comments before further processing.
        size_t commentPos = line.find("zZz");
        if (commentPos != std::string::npos) {
            line = line.substr(0, commentPos);
        }

        // Trim right whitespace
        line.erase(std::find_if(line.rbegin(), line.rend(), [](unsigned char ch) {
            return !std::isspace(ch);
        }).base(), line.end());

        if (line.empty()) continue;

        // Check indentation (Solar Beam Rule)
        // We count spaces to determine the indentation level.
        int indentCount = 0;
        bool hasTab = false;
        for (char c : line) {
            if (c == ' ') indentCount++;
            else if (c == '\t') { hasTab = true; break; }
            else break;
        }

        if (hasTab) throw std::runtime_error("Poison Type: Tab character detected");
        if (indentCount % 4 != 0) throw std::runtime_error("The attack missed!");

        int level = indentCount / 4;
        // Emit an INDENT token so the parser knows the nesting level of this line.
        tokens.push_back({TOKEN_INDENT, "", lineNum, level});

        std::string trimmedLine = trim(line);
        tokenizeLine(trimmedLine, lineNum);
    }
    
    tokens.push_back({TOKEN_EOF, "", lineNum, 0});
    return tokens;
}

// tokenizeLine
// Processes a single line after indentation has been handled.
// Identifies Section Headers or Key-Value pairs.
void Lexer::tokenizeLine(const std::string& line, int lineNum) {
    // Section Headers (Evolution Stages)
    // We check for specific patterns like (o) ... (o)
    if (startsWith(line, "(o) ") && endsWith(line, " (o)")) {
        tokens.push_back({TOKEN_SECTION_OPEN, "", lineNum, 1});
        std::string key = line.substr(4, line.length() - 8);
        tokens.push_back({TOKEN_IDENTIFIER, key, lineNum, 0});
        tokens.push_back({TOKEN_SECTION_CLOSE, "", lineNum, 1});
        return;
    }
    if (startsWith(line, "(O) ") && endsWith(line, " (O)")) {
        tokens.push_back({TOKEN_SECTION_OPEN, "", lineNum, 2});
        std::string key = line.substr(4, line.length() - 8);
        tokens.push_back({TOKEN_IDENTIFIER, key, lineNum, 0});
        tokens.push_back({TOKEN_SECTION_CLOSE, "", lineNum, 2});
        return;
    }
    if (startsWith(line, "(@) ") && endsWith(line, " (@)")) {
        tokens.push_back({TOKEN_SECTION_OPEN, "", lineNum, 3});
        std::string key = line.substr(4, line.length() - 8);
        tokens.push_back({TOKEN_IDENTIFIER, key, lineNum, 0});
        tokens.push_back({TOKEN_SECTION_CLOSE, "", lineNum, 3});
        return;
    }

    // Key-Value Pairs
    // We use Regex to capture the key, the vine whip, and the value.
    std::regex re("^([a-zA-Z0-9_]+)\\s*(~{1,}>)\\s*(.*)$");
    std::smatch matches;
    if (std::regex_match(line, matches, re)) {
        std::string key = matches[1];
        std::string valStr = matches[3];
        
        tokens.push_back({TOKEN_IDENTIFIER, key, lineNum, 0});
        tokens.push_back({TOKEN_VINE_WHIP, "", lineNum, 0});
        tokenizeValue(valStr, lineNum);
        return;
    }

    throw std::runtime_error("It hurt itself in its confusion!");
}

// tokenizeValue
// Parses the value part of a key-value pair.
void Lexer::tokenizeValue(const std::string& valStr, int lineNum) {
    std::string s = trim(valStr);
    if (s.empty()) return;

    // String Literal
    if (startsWith(s, "\"") && endsWith(s, "\"")) {
        tokens.push_back({TOKEN_STRING, s.substr(1, s.length() - 2), lineNum, 0});
        return;
    }
    // Boolean: SuperEffective (True)
    if (s == "SuperEffective") {
        tokens.push_back({TOKEN_BOOL, "true", lineNum, 0});
        return;
    }
    // Boolean: NotVeryEffective (False)
    if (s == "NotVeryEffective") {
        tokens.push_back({TOKEN_BOOL, "false", lineNum, 0});
        return;
    }
    // Null: MissingNo
    if (s == "MissingNo") {
        tokens.push_back({TOKEN_NULL, "", lineNum, 0});
        return;
    }
    // Array: <| ... |>
    if (startsWith(s, "<|") && endsWith(s, "|>")) {
        tokens.push_back({TOKEN_ARRAY_START, "", lineNum, 0});
        std::string inner = s.substr(2, s.length() - 4);
        std::stringstream ss(inner);
        std::string segment;
        bool first = true;
        while(std::getline(ss, segment, ',')) {
            if (!first) tokens.push_back({TOKEN_COMMA, "", lineNum, 0});
            tokenizeValue(segment, lineNum); // Recursive call for array elements
            first = false;
        }
        tokens.push_back({TOKEN_ARRAY_END, "", lineNum, 0});
        return;
    }

    // Number (Int/Float)
    try {
        size_t pos;
        std::stoi(s, &pos);
        if (pos == s.length()) {
            tokens.push_back({TOKEN_NUMBER, s, lineNum, 0});
            return;
        }
    } catch (...) {}

    try {
        size_t pos;
        std::stod(s, &pos);
        if (pos == s.length()) {
            tokens.push_back({TOKEN_NUMBER, s, lineNum, 0});
            return;
        }
    } catch (...) {}

    throw std::runtime_error("Target is immune!");
}

std::string Lexer::trim(const std::string& str) {
    size_t first = str.find_first_not_of(" \t");
    if (std::string::npos == first) return "";
    size_t last = str.find_last_not_of(" \t");
    return str.substr(first, (last - first + 1));
}

bool Lexer::startsWith(const std::string& str, const std::string& prefix) {
    return str.size() >= prefix.size() && str.compare(0, prefix.size(), prefix) == 0;
}

bool Lexer::endsWith(const std::string& str, const std::string& suffix) {
    return str.size() >= suffix.size() && str.compare(str.size() - suffix.size(), suffix.size(), suffix) == 0;
}

File: go-bson/lexer.go
package main

import (
	"bufio"
	"errors"
	"fmt"
	"regexp"
	"strings"
)

// TokenType represents the type of a token
// We use an integer enum for efficiency and type safety.
type TokenType int

const (
	TOKEN_HEADER        TokenType = iota // The "BULBA!" header
	TOKEN_INDENT                  // Value holds the level (0, 1, 2, 3)
	TOKEN_SECTION_OPEN            // (o), (O), (@) - Marks start of a section
	TOKEN_SECTION_CLOSE           // (o), (O), (@) - Marks end of a section header
	TOKEN_IDENTIFIER              // Keys (e.g., app_name, host)
	TOKEN_VINE_WHIP               // The assignment operator ~~~~>
	TOKEN_STRING                  // String literals "value"
	TOKEN_NUMBER                  // Numeric values 123, 4.5
	TOKEN_BOOL                    // Boolean values (SuperEffective, NotVeryEffective)
	TOKEN_NULL                    // Null value (MissingNo)
	TOKEN_ARRAY_START             // <|
	TOKEN_ARRAY_END               // |>
	TOKEN_COMMA                   // ,
	TOKEN_EOF                     // End of File marker
)

type Token struct {
	Type    TokenType
	Literal string // The actual text content of the token
	Line    int    // Line number for error reporting
	Level   int    // For INDENT and SECTION tokens, stores the nesting level
}

// Lexer performs lexical analysis on the input string.
// It reads the input line by line and converts it into a slice of Tokens.
// This separates the "what is this text?" logic from the "what does this structure mean?" logic.
func Lex(content string) ([]Token, error) {
	var tokens []Token
	scanner := bufio.NewScanner(strings.NewReader(content))
	lineNum := 0
	firstLine := true

	for scanner.Scan() {
		line := scanner.Text()
		lineNum++

		// Header check: The very first line must be the specific cry.
		if firstLine {
			if line != "BULBA!" {
				return nil, errors.New("Status: Fainted")
			}
			tokens = append(tokens, Token{Type: TOKEN_HEADER, Literal: "BULBA!", Line: lineNum})
			firstLine = false
			continue
		}

		// Handle Comments (Sleep Powder)
		// We strip out comments before further processing.
		if idx := strings.Index(line, "zZz"); idx != -1 {
			line = line[:idx]
		}

		// Check for tabs (Poison Type)
		// Tabs are strictly forbidden.
		if strings.Contains(line, "\t") {
			return nil, errors.New("Poison Type: Tab character detected")
		}

		// Trim right whitespace
		line = strings.TrimRight(line, " \r\n")
		if len(line) == 0 {
			continue
		}

		// Count Indentation (Solar Beam Rule)
		// We count spaces to determine the indentation level.
		indentCount := 0
		for _, char := range line {
			if char == ' ' {
				indentCount++
			} else {
				break
			}
		}

		if indentCount%4 != 0 {
			return nil, errors.New(ErrIndentation)
		}
		level := indentCount / 4
		// Emit an INDENT token so the parser knows the nesting level of this line.
		tokens = append(tokens, Token{Type: TOKEN_INDENT, Level: level, Line: lineNum})

		trimmedLine := strings.TrimSpace(line)

		// Tokenize the rest of the line
		err := tokenizeLine(&tokens, trimmedLine, lineNum)
		if err != nil {
			return nil, err
		}
	}

	tokens = append(tokens, Token{Type: TOKEN_EOF, Line: lineNum})
	return tokens, nil
}

// tokenizeLine processes a single line after indentation has been handled.
func tokenizeLine(tokens *[]Token, line string, lineNum int) error {
	// Check for Section Headers (Evolution Stages)
	// We look for patterns like (o) key (o)
	if strings.HasPrefix(line, "(o) ") && strings.HasSuffix(line, " (o)") {
		*tokens = append(*tokens, Token{Type: TOKEN_SECTION_OPEN, Level: 1, Line: lineNum})
		key := line[4 : len(line)-4]
		*tokens = append(*tokens, Token{Type: TOKEN_IDENTIFIER, Literal: key, Line: lineNum})
		*tokens = append(*tokens, Token{Type: TOKEN_SECTION_CLOSE, Level: 1, Line: lineNum})
		return nil
	}
	if strings.HasPrefix(line, "(O) ") && strings.HasSuffix(line, " (O)") {
		*tokens = append(*tokens, Token{Type: TOKEN_SECTION_OPEN, Level: 2, Line: lineNum})
		key := line[4 : len(line)-4]
		*tokens = append(*tokens, Token{Type: TOKEN_IDENTIFIER, Literal: key, Line: lineNum})
		*tokens = append(*tokens, Token{Type: TOKEN_SECTION_CLOSE, Level: 2, Line: lineNum})
		return nil
	}
	if strings.HasPrefix(line, "(@) ") && strings.HasSuffix(line, " (@)") {
		*tokens = append(*tokens, Token{Type: TOKEN_SECTION_OPEN, Level: 3, Line: lineNum})
		key := line[4 : len(line)-4]
		*tokens = append(*tokens, Token{Type: TOKEN_IDENTIFIER, Literal: key, Line: lineNum})
		*tokens = append(*tokens, Token{Type: TOKEN_SECTION_CLOSE, Level: 3, Line: lineNum})
		return nil
	}

	// Check for Key-Value Pairs
	// Regex: key ~~~~> value
	re := regexp.MustCompile(`^([a-zA-Z0-9_]+)\s*(~{1,}>)\s*(.*)$`)
	matches := re.FindStringSubmatch(line)
	if matches != nil {
		key := matches[1]
		// vine := matches[2]
		valStr := matches[3]

		*tokens = append(*tokens, Token{Type: TOKEN_IDENTIFIER, Literal: key, Line: lineNum})
		*tokens = append(*tokens, Token{Type: TOKEN_VINE_WHIP, Line: lineNum})

		return tokenizeValue(tokens, valStr, lineNum)
	}

	return errors.New(ErrSyntax)
}

// tokenizeValue parses the value part of a key-value pair.
func tokenizeValue(tokens *[]Token, valStr string, lineNum int) error {
	valStr = strings.TrimSpace(valStr)
	if valStr == "" {
		return nil
	}

	// String Literal
	if strings.HasPrefix(valStr, "\"") && strings.HasSuffix(valStr, "\"") {
		*tokens = append(*tokens, Token{Type: TOKEN_STRING, Literal: valStr[1 : len(valStr)-1], Line: lineNum})
		return nil
	}

	// Boolean: SuperEffective (True)
	if valStr == "SuperEffective" {
		*tokens = append(*tokens, Token{Type: TOKEN_BOOL, Literal: "true", Line: lineNum})
		return nil
	}
	// Boolean: NotVeryEffective (False)
	if valStr == "NotVeryEffective" {
		*tokens = append(*tokens, Token{Type: TOKEN_BOOL, Literal: "false", Line: lineNum})
		return nil
	}

	// Null: MissingNo
	if valStr == "MissingNo" {
		*tokens = append(*tokens, Token{Type: TOKEN_NULL, Line: lineNum})
		return nil
	}

	// Array: <| ... |>
	if strings.HasPrefix(valStr, "<|") && strings.HasSuffix(valStr, "|>") {
		*tokens = append(*tokens, Token{Type: TOKEN_ARRAY_START, Line: lineNum})
		inner := strings.TrimSpace(valStr[2 : len(valStr)-2])
		if inner != "" {
			parts := strings.Split(inner, ",")
			for i, p := range parts {
				if i > 0 {
					*tokens = append(*tokens, Token{Type: TOKEN_COMMA, Line: lineNum})
				}
				// Recursive call for array elements
				if err := tokenizeValue(tokens, strings.TrimSpace(p), lineNum); err != nil {
					return err
				}
			}
		}
		*tokens = append(*tokens, Token{Type: TOKEN_ARRAY_END, Line: lineNum})
		return nil
	}

	// Number (Int/Float)
	// Simple check: if it looks like a number
	if _, err := fmt.Sscan(valStr, new(float64)); err == nil {
		*tokens = append(*tokens, Token{Type: TOKEN_NUMBER, Literal: valStr, Line: lineNum})
		return nil
	}

	return errors.New(ErrType)
}
